opentelemetry-sdk/src/opentelemetry/sdk/trace/sampling.py

```diff
@@ -372,26 +407,52 @@ def __init__(self, rate: float):
 
 
 def _get_from_env_or_default() -> Sampler:
-    trace_sampler = os.getenv(
+    trace_sampler_name = os.getenv(
         OTEL_TRACES_SAMPLER, "parentbased_always_on"
     ).lower()
-    if trace_sampler not in _KNOWN_SAMPLERS:
-        _logger.warning("Couldn't recognize sampler %s.", trace_sampler)
-        trace_sampler = "parentbased_always_on"
 
-    if trace_sampler in ("traceidratio", "parentbased_traceidratio"):
+    if trace_sampler_name in _KNOWN_SAMPLERS:
+        if trace_sampler_name in ("traceidratio", "parentbased_traceidratio"):
+            try:
+                rate = float(os.getenv(OTEL_TRACES_SAMPLER_ARG))
+            except ValueError:
+                _logger.warning(
+                    "Could not convert TRACES_SAMPLER_ARG to float."
+                )
+                rate = 1.0
+            return _KNOWN_SAMPLERS[trace_sampler_name](rate)
+        return _KNOWN_SAMPLERS[trace_sampler_name]
+    else:
         try:
-            rate = float(os.getenv(OTEL_TRACES_SAMPLER_ARG))
-        except ValueError:
-            _logger.warning("Could not convert TRACES_SAMPLER_ARG to float.")
-            rate = 1.0
-        return _KNOWN_SAMPLERS[trace_sampler](rate)
-
-    return _KNOWN_SAMPLERS[trace_sampler]
+            trace_sampler_factory = _import_sampler_factory(trace_sampler_name)
+            sampler_arg = os.getenv(OTEL_TRACES_SAMPLER_ARG, "")
+            trace_sampler = trace_sampler_factory(sampler_arg)
+            if not isinstance(trace_sampler, Sampler):
+                message = (
+                    "Output of traces sampler factory, %s, was not a Sampler object."
+                    % trace_sampler_factory
+                )
+                _logger.warning(message)
+                raise ValueError(message)
+            return trace_sampler
+        except Exception as exc:
+            _logger.warning(
+                "Failed to initialize custom sampler, %s: %s",
+                trace_sampler_name,
+                exc,
+            )
+            return _KNOWN_SAMPLERS["parentbased_always_on"]
 
 
 def _get_parent_trace_state(parent_context) -> Optional["TraceState"]:
     parent_span_context = get_current_span(parent_context).get_span_context()
     if parent_span_context is None or not parent_span_context.is_valid:
         return None
     return parent_span_context.trace_state
+
+
+def _import_sampler_factory(sampler_name: str) -> Callable[[str], Sampler]:
```

Comment:
Having logic related to importing and defaulting isn't really related to the component itself, so it would make sense to move this and `_get_from_env_or_default` to a separate file under `_configuration` probably. Would make it easier to read once we start using factories to convert all of our components. Not a blocker for this PR but probably something to keep in mind for the future.

Reply:
This would also create a circular dependency issue

Reply:
@srikanthccv Are you okay with these changes?